xpad.c

```diff
@@ -89,6 +90,12 @@
 #define XTYPE_XBOXONE     3
 #define XTYPE_UNKNOWN     4
 
+#define PTYPE_STD         0
+#define PTYPE_EONE        1
+#define PTYPE_ETWOOLD     2
+#define PTYPE_ETWO5       3
+#define PTYPE_ETWO5DOT11  4
```

Comment:
Is this convention OK for differentiating packet formats?

Reply:
I dont get the naming scheme. This is about the paddle packet. So I would suggest:

```
PKT_PADDLES_NONE
PKT_PADDLES_1_0
PKT_PADDLES_4_0
PKT_PADDLES_5_0
PKT_PADDLES_5_11
```

Reply:
I tried to keep the same shorter style of the XTYPE variables, but I'm quite happy to be more verbose.

There are two things that I thing are important that I want to highlight, though:
- This should describe the packet type generally, rather than being something specific to paddles. This is why I used a value of "STD" ("Standard") rather than "NONE". I believe that this will keep the driver more maintainable while also proving useful for future applications, such as exposing the configuration interface.
- Individual values should not be directly interchangeable with "firmware version", because this is not going to be consistent across device generations and manufacturers. This is an issue even now, as "1_0" doesn't correspond to an actual firmware number -- my "EONE" simply means "The Xbox **E**lite Series **One**", as that controller has its own special format.

Given these points, here's my new suggestion:

```
PKT_STANDARD
PKT_ELITE1
PKT_ELITE2_FW_OLD
PKT_ELITE2_FW_5_EARLY
PKT_ELITE2_FW_5_11
```

Reply:
In xpadneo we use XB1S, XBE2 and XBXS for the controller names, at least in the source code. xpad could do the same and add XBE1, XB360 and the likes. In xpadneo, I version the firmware packet formats independently of the controller firmware version when the format changes in an incompatible way (i.e. conflict bits), so we have packet format v1 for the early (which looks like a broken format, partially overwriting and duplicating itself shifted by some bytes, thus moving some overwritten bits to an awkward position, at least in HID), and v2 for the later formats (which only added some new bits). After that, there's just some bounds checking for not trying to access bits beyond the packet size.

In this sense, `XB` acts as a prefix identifying a model. Something similar could be done for the firmware / packet format with a `FW` prefix, e.g. `PKT_XBE2_FW511`.

Reply:
For the obviously broken packet format, I'll emit a kernel warning recommending to upgrade the firmware - so we can easily detect that in user submitted logs... Who knows what other surprises a hidden in that broken format.

Reply:
I like this shorthand. In that case, how's about this?

```
PKT_XB
PKT_XBE1
PKT_XBE2_FW_OLD
PKT_XBE2_FW_5_EARLY
PKT_XBE2_FW_5_11
```

I'm fine with this format as well, which feels more regular, but less readable:

```
PKT_XB
PKT_XBE1
PKT_XBE2_FWOLD
PKT_XBE2_FW5EARLY
PKT_XBE2_FW511
```

At the end of the day, I am probably bikeshedding a little excessively here, since these enum-like values can be changed later without much issue.

Reply:
> For the obviously broken packet format, I'll emit a kernel warning recommending to upgrade the firmware - so we can easily detect that in user submitted logs... Who knows what other surprises a hidden in that broken format.

@paroj: What's the xpad policy on emitting firmware related warnings? We may want to do something similar when we detect an outdated packet type

On that note: I _am_ curious to see what the USB packets for that early firmware looked like. If that weird duplication isn't a BT-only bug, I want to document and implement it in the driver for the sake of completeness.

Reply:
Well, at least as part of the bug reporting process in xpadneo, I'm asking for dmesg output. On that note, I've never seen the outdated packet format warning in the wild - so that firmware is probably not actually used very much. Many users also use their controllers with an Xbox and automatically receive updates there. That outdated/broken firmware may very well have gone extinct by now. Yes, there's `dev_dbg` but given the nature of this problem, I've chosen to write a warning instead just once per module load to make it a little bit more visible without spamming the log.